controller/konnect/ops/ops_kongconsumer.go

```diff
@@ -118,6 +141,189 @@ func updateConsumer(
+// - sets Programmed to False if any operation fails
+// - sets KongConsumerGroupRefsValid to False if any of the ConsumerGroup references are invalid
+// - sets KongConsumerGroupRefsValid to True if all ConsumerGroup references are valid
+// It returns an error if any of the operations fail (fetching KongConsumers from cache, fetching the actual Konnect
+// ConsumerGroups the Consumer is assigned to, etc.).
+func handleConsumerGroupAssignments(
+	ctx context.Context,
+	consumer *configurationv1.KongConsumer,
+	cl client.Client,
+	cgSDK ConsumerGroupSDK,
+	cpID string,
+) error {
+	// Resolve the Konnect IDs of the ConsumerGroups referenced by the KongConsumer.
+	desiredConsumerGroupsIDs, invalidConsumerGroups, err := resolveConsumerGroupsKonnectIDs(ctx, consumer, cl)
+	if err != nil {
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				conditions.KonnectEntityProgrammedConditionType,
+				metav1.ConditionFalse,
+				conditions.KonnectEntityProgrammedReasonFailedToResolveConsumerGroupRefs,
+				err.Error(),
+				consumer.GetGeneration(),
+			),
+			consumer,
+		)
+		return err
+	}
+
+	// Even if we have invalid ConsumerGroup references, we carry on with the ones that are valid. Invalid ones will be
+	// reported in the condition.
+	populateConsumerGroupRefsValidCondition(invalidConsumerGroups, consumer)
+
+	// Reconcile the ConsumerGroups assigned to the KongConsumer in Konnect (list the actual ConsumerGroups, calculate the
+	// difference, and add/remove the Consumer from the ConsumerGroups accordingly).
+	if err := reconcileConsumerGroupsWithKonnect(ctx, desiredConsumerGroupsIDs, cgSDK, cpID, consumer); err != nil {
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				conditions.KonnectEntityProgrammedConditionType,
+				metav1.ConditionFalse,
+				conditions.KonnectEntityProgrammedReasonFailedToReconcileConsumerGroupsWithKonnect,
+				err.Error(),
+				consumer.GetGeneration(),
+			),
+			consumer,
+		)
+		return err
+	}
+	return nil
+}
+
+// reconcileConsumerGroupsWithKonnect reconciles the ConsumerGroups assigned to a KongConsumer in Konnect. It calculates
+// the difference between the desired ConsumerGroups and the actual ConsumerGroups in Konnect and adds or removes the
+// Consumer from the ConsumerGroups accordingly. It returns an error if any of the Konnect operations fail.
+func reconcileConsumerGroupsWithKonnect(
+	ctx context.Context,
+	desiredConsumerGroupsIDs []string,
+	cgSDK ConsumerGroupSDK,
+	cpID string,
+	consumer *configurationv1.KongConsumer,
+) error {
+	// List the ConsumerGroups that the Consumer is assigned to in Konnect.
+	cgsResp, err := cgSDK.ListConsumerGroupsForConsumer(ctx, sdkkonnectops.ListConsumerGroupsForConsumerRequest{
+		ControlPlaneID: cpID,
+		ConsumerID:     consumer.GetKonnectStatus().GetKonnectID(),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list ConsumerGroups for Consumer %s/%s: %w", consumer.GetNamespace(), consumer.GetName(), err)
```

Comment:
Instead of getting the ns and name you can use 

```suggestion
		return fmt.Errorf("failed to list ConsumerGroups for Consumer %s: %w", client.ObjectKeyFromObject(consumer), err)
```